aws/hll/aws-s3-transfer-manager/src/download/body.rs

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+use crate::download::worker::ChunkResponse;
+use crate::error::TransferError;
+use aws_smithy_types::byte_stream::AggregatedBytes;
+use std::cmp;
+use std::cmp::Ordering;
+use std::collections::BinaryHeap;
+use tokio::sync::mpsc;
+
+/// Stream of binary data representing an object's contents.
+///
+/// Wraps potentially multiple streams of binary data into a single coherent stream.
+/// The data on this stream is sequenced into the correct order.
+#[derive(Debug)]
+pub struct Body {
+    inner: UnorderedBody,
+    sequencer: Sequencer,
+}
+
+type BodyChannel = mpsc::Receiver<Result<ChunkResponse, TransferError>>;
+
+impl Body {
+    /// Create a new empty Body
+    pub fn empty() -> Self {
+        Self::new_from_channel(None)
+    }
+
+    pub(crate) fn new(chunks: BodyChannel) -> Self {
+        Self::new_from_channel(Some(chunks))
+    }
+
+    fn new_from_channel(chunks: Option<BodyChannel>) -> Self {
+        Self {
+            inner: UnorderedBody::new(chunks),
+            sequencer: Sequencer::new(),
+        }
+    }
+
+    /// Convert this body into an unordered stream of chunks.
+    pub(crate) fn unordered(self) -> UnorderedBody {
+        self.inner
+    }
+
+    /// Pull the next chunk of data off the stream.
+    ///
+    /// Returns [None] when there is no more data.
+    /// Chunks returned from a [Body] are guaranteed to be sequenced
+    /// in the right order.
+    pub async fn next(&mut self) -> Option<Result<AggregatedBytes, TransferError>> {
```

Comment:
Seems weird to return an optional result. Would this make more sense?
```suggestion
    pub async fn next(&mut self) -> Result<Option<AggregatedBytes>, TransferError> {
```

Reply:
I would say it's not so strange in Rust for a iteration method to return an option

Reply:
Exactly, this is commonly used for iteration.